### PR TITLE
Add Support for Protein Section, Peptide Section and PSM Section

### PR DESCRIPTION
**Pull request by epoyraz**

### CODE REVIEW

**Comment by timosachsenberg** [comment]
Hi,
thanks for the pull request. Overall it looks good - I just have some minor comments and questions:

One thing I didn't undestand is the function: private DataTableSpec createDummySpec(String HeaderType)
Why is this needed?
Also it seems as if column types are created by default as string for every entry. Maybe here, a mapping to known integer or double types could be done for the standard columns (the ones defined in the spec document)

**Comment by timosachsenberg** [comment]
Took another look and I think I understand now:
You determine the cell types based on the column names - which is fine. Right now you match  only the column names for small molecules so the column names in the other sections.

**Comment by jpfeuffer** [comment]
I have the same question as Timo about the createDummySpec function. I am asking because this function limits us to a predefined set of columns (which doesnt always fit all optional opt_ID columns).
This is an honest question because I do not know if this is a limitation on the KNIME side. But it would certainly be an improvement.
Anyway: nice work, thanks!

File: de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java [hunk 1]
```diff
@@ -85,7 +85,7 @@
     private static final NodeLogger logger = NodeLogger
             .getLogger(SmallMoleculeMzTabReaderNodeModel.class);
 
-    private int metaDataRowIdx, smallMolRowIdx;
+    private int metaDataRowIdx,proteinRowIdx,peptideRowIdx,psmRowIdx, smallMolRowIdx;
```
Comment (aiche) [inline de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java]: please follow the general coding style and put a space after each comma

File: de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java [hunk 2]
```diff
@@ -106,10 +106,18 @@ private DataTableSpec createMetaDataSectionSpec() {
...
     protected MzTabReaderNodeModel() {
         super(new PortType[] { new PortType(IURIPortObject.class) },
-                new PortType[] { new PortType(BufferedDataTable.class),
-                        new PortType(BufferedDataTable.class) });
+                new PortType[] { 
+        		new PortType(BufferedDataTable.class),
+        		new PortType(BufferedDataTable.class),
+        		new PortType(BufferedDataTable.class),
+        		new PortType(BufferedDataTable.class),
+        		new PortType(BufferedDataTable.class),
```
Comment (aiche) [inline de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java]: please remove the superfluous comma

File: de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java [hunk 3]
```diff
@@ -106,10 +106,18 @@ private DataTableSpec createMetaDataSectionSpec() {
...
+        });
```
Comment (aiche) [inline de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java]: indent looks weird

**Comment by epoyraz** [comment]
@timosachsenberg : yeah you are right. I'll fix the type issue for the other column types. 

Hi,

the Reason i used createDummySpec was : 
Not all Section are present in a mzTab File and i am not allowed to return "null" inside a BufferedDataTable/Outport. Therefore i am crating an empty (0 Rows) dummy Table, only for the Sections that are not present in the file. I agree, it's rather a workaround. Open for suggestions :)

File: de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java [hunk 4]
```diff
@@ -154,10 +176,10 @@ protected MzTabReaderNodeModel() {
...
             // validate mzTab starting point
             String line = brReader.readLine();
-            if (!line.trim().startsWith("MTD")) {
-                throw new InvalidMzTabFormatException(
-                        "Invalid start of file: mzTab file should start with the line: 'MTD\tmzTab-version\t1.0.0'");
-            }
+//            if (!line.trim().startsWith("MTD")) {
+//                throw new InvalidMzTabFormatException(
+//                        "Invalid start of file: mzTab file should start with the line: 'MTD\tmzTab-version\t1.0.0'");
+//            }
```
Comment (aiche) [inline de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java]: please remove unused code

File: de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java [hunk 5]
```diff
@@ -188,11 +225,25 @@ protected MzTabReaderNodeModel() {
                 // allow knime to cancel node execution
                 exec.checkCanceled();
             } while ((line = brReader.readLine()) != null);
+                        
+            if (proteinDataContainer==null) proteinDataContainer = exec.createDataContainer(createDummySpec("PSH"));
+            if (peptideDataContainer==null) peptideDataContainer =  exec.createDataContainer(createDummySpec("PEH"));
+            if (psmDataContainer==null)  psmDataContainer = exec.createDataContainer(createDummySpec("PSH"));
+            if (smallMolContainer==null) smallMolContainer =  exec.createDataContainer(createDummySpec("SMH"));
```
Comment (aiche) [inline de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java]: code style [spaces around operator]

File: de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java [hunk 6]
```diff
@@ -203,45 +254,116 @@ protected MzTabReaderNodeModel() {
                 brReader.close();
         }
 
-        return new BufferedDataTable[] { metaDataTable, smallMolTable };
+        return new BufferedDataTable[] { metaDataTable, proteinDataTable,peptideDataTable,psmDataTable,smallMolTable };
```
Comment (aiche) [inline de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java]: style

File: de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java [hunk 7]
```diff
@@ -276,8 +398,28 @@ private void parseMTDLine(BufferedDataContainer metaDataContainer,
...
+        for (int i = 1; i < line_entries.length; ++i) {
+            DataType type = getDataType(line_entries[i]);
+            colSpecs[i - 1] = new DataColumnSpecCreator(line_entries[i], type)
+                    .createSpec();
+        }
+
+        return new DataTableSpec(colSpecs);
+    }
+    
+    private DataTableSpec createDummySpec(String HeaderType) {
```
Comment (aiche) [inline de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java]: please document the intention of this method and ideally give it a more meaningful name.

File: de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java [hunk 8]
```diff
@@ -358,7 +502,7 @@ protected void reset() {
     @Override
     protected DataTableSpec[] configure(final PortObjectSpec[] inSpecs)
             throws InvalidSettingsException {
-        return new DataTableSpec[] { createMetaDataSectionSpec(), null };
+        return new DataTableSpec[] { createMetaDataSectionSpec(), null,null,null,null };
```
Comment (aiche) [inline de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java]: style

File: de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java [hunk 9]
```diff
@@ -276,8 +398,28 @@ private void parseMTDLine(BufferedDataContainer metaDataContainer,
...
+        if(HeaderType.equals("PSH")) line = "PSH	sequence	PSM_ID	accession	unique	database	database_version	search_engine	search_engine_score[1]	modifications	spectra_ref	retention_time	charge	exp_mass_to_charge	calc_mass_to_charge	pre	post	start	end ";
+        if(HeaderType.equals("SMH")) line = "SMH	identifier	chemical_formula	smiles	inchi_key	description	exp_mass_to_charge	calc_mass_to_charge	charge	retention_time	taxid	species	database	database_version	spectra_ref	search_engine	best_search_engine_score[1]	modifications" ;
```
Comment (aiche) [inline de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java]: please reduce the number of columns to the minimum required set of mzTab.

Comment (aiche) [inline de.openms.knime.mzTab/src/de/openms/knime/mztab/MzTabReaderNodeModel.java]: thinking a second time about this I would suggest to directly initialize the columnspecs or at least the array of strings. Having those long tab separated lines in the code looks weird and doesn't feel right

**Comment by aiche** [comment]
apart from the above comments, thanks for taking the time to extend the mzTab reader

**Comment by epoyraz** [comment]
Thanks for the input, i'll address all of these issues in the next couple of Days.

**Comment by timosachsenberg** [comment]
"Not all Section are present in a mzTab File and i am not allowed to return "null" inside a BufferedDataTable/Outport. Therefore i am crating an empty (0 Rows) dummy Table, "

@aiche do you know if there is a way to return empty tables?
Returning dummy headers is not ideal as the mandatory columns already depend on the type of mzTab file and also e.g. the number of score columns? Easier would be to just return an empty one.

**Comment by aiche** [comment]
Off the top of my head I would agree with @epoyraz with respect to returning empty tables. Maybe you could ask Thorsten or Bernd from KNIME if they have an idea.

**Comment by timosachsenberg** [comment]
Ok will do. Maybe @AlexanderFillbrunn is also watching...

**Comment by timosachsenberg** [comment]
+1 looks good 
@jpfeuffer will this automatically show up in our nightly KNIME builds?

**Comment by jpfeuffer** [comment]
It should. If not, reply here and I will try to fix it.

**Comment by timosachsenberg** [comment]
Hi @epoyraz  can you take a look at https://github.com/genericworkflownodes/de.openms.knime/issues/10
Maybe some column types need to be adapted

**Comment by timosachsenberg** [comment]
Julianus fixed it

**Pull request merged**